### PR TITLE
Implemented contracts API

### PR DESCRIPTION
**Pull request by fvukelic**
Implemented endpoints for contracts. This also covers all the basic tests for the marketplace with an Sqlite database. 

- TG-71
- TG-72
- TG-73
- TG-86

### CODE REVIEW

File: flocx_market/tests/system/test_offer.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
+from flocx_market.db.sqlalchemy.offer_api import OfferApi
```
Comment (larsks) [inline flocx_market/tests/system/test_offer.py]: This doesn't seem to be related to implementing the contract api. Maybe open an issue for adding new tests and then add it in a new pull request?

File: flocx_market/tests/unit/models/test_offer_unit.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+from flocx_market.tests.unit.unit_base_test import UnitBaseTest
```
Comment (larsks) [inline flocx_market/tests/unit/models/test_offer_unit.py]: Ditto on this one.

File: flocx_market/tests/integration/models/test_offer_integration.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+from flocx_market.db.sqlalchemy.offer_api import OfferApi
```
Comment (larsks) [inline flocx_market/tests/integration/models/test_offer_integration.py]: ...and this one too.

File: flocx_market/db/sqlalchemy/contract_api.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    def as_dict(self):
+        d = {}
+        for col in self.__table__.columns:
+            if col.name in ['time_created', 'start_time', 'end_time']:
+                d[col.name] = getattr(self, col.name).isoformat()
+            else:
+                d[col.name] = getattr(self, col.name)
+        offer_model_objects = [offer for offer in self.offers.all()]
+
+        # needed to avoid foreign key constraint issue since model and api are separated
```
[the remark this answers is not in the document]
Comment (tzumainn) [reply flocx_market/db/sqlalchemy/contract_api.py]: I agree; this potentially makes the function very expensive database-wise.

File: flocx_market/db/sqlalchemy/models.py [hunk 5]
```diff
@@ -65,3 +66,48 @@ class Offer(Base):
...
+    def contract_to_dict(self):
+        d = {}
+        for col in self.__table__.columns:
+            val = getattr(self, col.name)
+            if type(val) == datetime.datetime:
+                d[col.name] = val.isoformat()
+            else:
+                d[col.name] = val
+
+        offer_model_objects = [offer for offer in self.offers.all()]
```
Comment (tzumainn) [inline flocx_market/db/sqlalchemy/models.py]: This seems pretty dangerous; you're iterating through every single offer in the database, and if you're fetching, say, all contracts, you're doing that for every single contract.

I don't think this expansion needs to be done at all; just print the offer_id. I'd suggest removing the `contract_to_dict` function, and just using `to_dict`.

Comment (fvukelic) [reply flocx_market/db/sqlalchemy/models.py]: Hey @tzumainn, yes, sure I can make those separate commits
I call self.offers.all(), that is only iterating through all the offers related to this contract, not all in the database. Should I expand the basic to_dict method or how do you think I should approach returning of offer_ids related to this contract?

Comment (tzumainn) [reply flocx_market/db/sqlalchemy/models.py]: Whoops, you're right!

I don't think you should expand the offer at all here; the current to_dict method is sufficient for simply having the value for offer_id, correct?

Comment (fvukelic) [reply flocx_market/db/sqlalchemy/models.py]: I'm not storing the offers in the contracts because there can be many offers that match 1 bid and we don't know the number. So instead I just keep a foreign key on offers for contracts and that's why I need to pull them with self.offers. What do you think it's the best solution in that case @tzumainn?

Comment (tzumainn) [reply flocx_market/db/sqlalchemy/models.py]: Ah - I think there may be a misunderstanding here. We shouldn't need to keep track of all the offers that match a single bid; the matcher will assign one of those offers to that bid, and create a contract.

It's possible that we haven't gotten the data model quite right yet, so if we change it in the future, that's fine. Either way, I think we just want to use a generic `to_dict` function to return to users through the REST API. Don't worry about following relationships or anything yet.

Comment (fvukelic) [reply flocx_market/db/sqlalchemy/models.py]: hmm okay, I'm not keeping track of them, I am receiving a list of offers from the matcher when a contract is being made, and then I update those offers to reference the contract that is passed in. But I can't keep the offers in the contract table since there can be many offers to one contract. In the API documentation it says that the contracts should return the list of offers that it is associated with, should I just leave that out then @tzumainn?

Comment (tzumainn) [reply flocx_market/db/sqlalchemy/models.py]: I may have misunderstood; I thought you meant offers matching bids, but you mean offers matching contracts. But I thought that if a bid is looking for 5 nodes, then 5 separate contracts are created for each node - in other words, each contract specifies exactly one node?

Comment (fvukelic) [reply flocx_market/db/sqlalchemy/models.py]: yes, sorry I was unclear there. Hmm, not sure, I was following the API documentation, there the contracts can have multiple offers and 1 bid only @tzumainn

Comment (tzumainn) [reply flocx_market/db/sqlalchemy/models.py]: Hm, interesting, maybe I'm wrong! In any case, I think my general point stands: the to_dict method here should just change the database object into a dict; it shouldn't do any additional object expansion.

**Comment by tzumainn** [comment]
I'm not entirely sure about the data model, but I think that's more me getting confused than anything else. I still don't think an offer should tie specifically into one contract, because it seems to me that an offer should be able to handle multiple contracts over its lifespan - but I know that's a point of contention.

In any case, I'm fine with this going in as-is, and possibly adjusting the data model later if needed. @larsks what do you think?

File: flocx_market/tests/unit/api/test_app_contract.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+                          start_time="2016-07-16T19:20:30",
+                          end_time="2016-07-16T19:20:30",
+                          cost=0.0,
+                          bid_id='test_bid_2',
+                          offers=[contract_1_offer.marketplace_offer_id]
+                          )
+
+
+@mock.patch('flocx_market.db.sqlalchemy.api.contract_get_all')
+def test_get_bids(mock_get_all, client):
```
Comment (larsks) [inline flocx_market/tests/unit/api/test_app_contract.py]: This should probably be `test_get_contracts` :)

**Comment by tzumainn** [comment]
Ah, we have to add `contract_id` to offers in a few places:

* flocx_market/objects/offer.py - add it to the fields
* flocx_market/tests/unit/api/test_app_offer.py, flocx_market/tests/unit/objects/test_object_offer.py - add contract_id to the test offers


*

**Comment by tzumainn** [comment]
Whoops, you did update objects/offer.py. It possibly has to have 'nullable=True'

**Comment by fvukelic** [comment]
@tzumainn had to add it to a few places, just now saw that the tests were failing :D they passed now

**Comment by tzumainn** [comment]
@fvukelic awesome, thanks! Looks good - merging now!
